src/logEntryProcessing.ts

```diff
@@ -1,5 +1,139 @@
+import { CloudWatchLogsLogEvent } from 'aws-lambda';
+
 export function isRequestLogEntry(line: string): boolean {
     return line.startsWith('START RequestId: ') ||
         line.startsWith('END RequestId: ') ||
         line.startsWith('REPORT RequestId: ')
+}
+
+export function fieldValue(text: string, fieldNameText: string, valueLength?: number): string {
+    const colonAndSpaceSeparator = 2;
+    return text.substr(text.indexOf(fieldNameText)+fieldNameText.length + colonAndSpaceSeparator, valueLength).trim();
+}
+
+/**
+ * Parse an AWS lambda report field into a field name and a value
+ * @param rawField The raw field that looks something like "Field Name: value unit"
+ */
+export function parseReportField(rawField: string): [string, any] {
+    const [rawFieldName, rawFieldValue] = rawField.split(':').map(s => s.trim());
+
+    const fieldNameNoSpaces = rawFieldName.replace(/ /g, '');
+    const fieldName = fieldNameNoSpaces.charAt(0).toLowerCase() + fieldNameNoSpaces.slice(1);
+
+    const [value, unit] = rawFieldValue.split(' ');
+
+    if (unit == 'ms' || unit == 'MB') {
+        // value should be numeric
+        const numericValue = parseFloat(value);
+        return [fieldName + unit, numericValue] as [string, any];
+    } else {
+        // we didn't recognise the unit, perhaps there isn't one
+        return [fieldName, rawFieldValue] as [string, any];
+    }
+}
+
+export function lambdaRequestLogData(line: string): StructuredLogData | undefined {
+    if (isRequestLogEntry(line)) {
+        const eventName = line.substr(0, line.indexOf(' '));
+        const requestId = fieldValue(line, 'RequestId', 36);
+        const base = {
+            lambdaEvent: eventName,
+            lambdaRequestId: requestId
+        };
+        let stats: StructuredLogData = {};
+        switch(eventName) {
+            case 'END':
+                // no other fields
+                break;
+            case 'START':
+                // extract Version:
+                const version = fieldValue(line, 'Version');
+                stats = {
+                    lambdaVersion: version
+                };
+                break;
+            case 'REPORT':
+                // extract other fields (conveniently tab separated)
+                const rawFields = line.split('\t').slice(1).map(s => s.trim()).filter(s => s.length > 0);
+                const fields: [string, any][] = rawFields.map(rawField => parseReportField(rawField));
+                stats = fields.reduce((acc: StructuredLogData, field) => {
+                    const [fieldName, fieldValue] = field;
+                    acc[fieldName] = fieldValue;
+                    return acc;
+                }, {});
+                break;
+        }
+
+        return Object.assign(base, {
+            lambdaStats: stats
+        });
+    } else {
+        return undefined;
```

Comment:
minor: could we invert the `if` condition to create an early return as it's slightly easier to read, imo.

```typescript
if (!isRequestLogEntry(line)) {
  return undefined;
}

// code on the happy path
```

Reply:
Sure: https://github.com/guardian/cloudwatch-logs-management/pull/44/commits/0ad4bdc301df54b508ea8af8b00e4ec72dbc138a